Changes allowing old Microsoft C Compilers still to build OCaml

dra27 opened the pull request:
The product of some testing of very old Microsoft C Compilers - posting the PR to ask very quickly for help with the very weird patch in https://github.com/ocaml/ocaml/commit/47d6ce213c91e86e223a64190a45286da1c590a5!

Each commit fixes three distinct faults:
- https://github.com/ocaml/ocaml/commit/b9ca576c7958eee33b82bac9a4cf89201994b249 (see also [MPR4887](http://caml.inria.fr/mantis/view.php?id=4887)) more correctly bounds the code for _set_invalid_parameter_handler with the proper detection for the secure API changes.
- https://github.com/ocaml/ocaml/commit/9f310fa0b02ce5f605bf18da600973bd3e00b466 fixes the lack of support for variadic macros in all editions of the C pre-processor prior to Visual Studio 2008 (I haven't bothered to correct the `DEBUG` branch for compilers that old - if someone ever has the need to debug `otherlibs/win32unix/select.c` on such a compiler, they're welcome to write the tedious macro to support it!)
- https://github.com/ocaml/ocaml/commit/47d6ce213c91e86e223a64190a45286da1c590a5 is necessary for Unix.in_channel_of_descr to work when compiled with Visual Studio .NET 2003. I freely admit that this compiler is so old as to be almost not worth any attention, but I was wondering if an expert looking at the simple patch (it was identified as a result of inserting `printf`s to identify the failure and then going from there...) may lead to a more proper patch with comments. It's pretty cool if OCaml does still build successfully on a compiler released 12 years ago - it also has the "benefit" of meaning that the MSVC port can be built using the same runtime as the MinGW port (i.e. you "need" Visual Studio .NET 2003 to build against msvcrt.dll)

Or the whole thing can be dropped - the whole exercise is part of revising the Windows build instructions and seeing just how far back support still extends. I personally find it interesting/impressive that so few changes are apparently necessary...

dra27 committed:
Correctly detect _set_invalid_parameter_handler

There is still at least one Microsoft-supported version of the CRT which
does not include the secure versions of various functions, and
consequently does not include `_set_invalid_parameter_handler`.

Tests for the definition and inclusion of
caml_install_invalid_parameter_handler refined to detect
`__STDC_SECURE_LIB__` defined in `crtdefs.h`. This is a badly documented,
but standardised, define. Although `__STDC_SECURE_LIB__` is compatible
with MinGW, I have retained the test for `_MSC_VER` and so kept this as an
MSVC-only patch.

dra27 committed:
Fix DEBUG_PRINT for older MS C Compilers

Support was added for variadic macros in the RTM version of Visual Studio
2005 (CL Version 14), but there are variants of the version 14 compiler
included in SDKs released before Visual Studio 2005 itself which do not
support them. Fix the non-DEBUG version of win32unix to compile correctly
and not display a warning for DEBUG_PRINT calls in
`otherlibs/win32unix/select.c`.

dra27 committed:
For discussion! Enable compilation using Visual C++ 2002 & 2003

This bizarre patch results in a full test-suite pass for Visual Studio
.NET 2002 & 2003. Without it, the following code segfaults (resulting in
6 failing tests):

```ocaml
  #load "unix.cma";;
  let (ifd, _) = Unix.pipe ();;
  Unix.in_channel_of_descr ifd;;
```

dra27 commented:
A quick test, as I happen to have a VM with Visual Studio 6 on it, confirms that Visual Studio .NET 2003 is the oldest Microsoft C Compiler which can compile OCaml (given these patches) - Visual C++ 6 can't even build ocamlrun :smile:

gasche commented:
I had a look a the proposed patches: they look safe (low regression risk).

dra27 commented:
The one which worries me is whatever the extra `fflush(NULL)` call fixes, because I have no idea why it fixes it! It feels very weird to be doing that for newer MSVC's where that's not needed.

dra27 commented:
Actually, having just read some more, it would definitely have to change - I assumed `fflush(NULL)` would be an erroneous call (effectively a no-op) but having RTFM'd, the actual behaviour of flushing all buffers is really not what I meant! However, `fflush(stdin)` also "works"...

dra27 commented:
One bug fix, one update in the light of some further testing:
- The DEBUG_PRINT fix test was incorrect and was breaking mingw - added `defined(_MSC_VER) &&` which fixes that :blush:
- As I was there, updated the compilation for Visual C++ 2003 to "flush" `stdin` and also added a guard around it so that it's only used for versions of Visual Studio older than 2005, but this still isn't a patch that should be committed...

oandrieu commented on asmrun/startup.c:
```diff
@@ -86,7 +86,7 @@ extern value caml_start_program (void);
 extern void caml_init_signals (void);
 
-#ifdef _MSC_VER
+#if defined(_MSC_VER) && __STDC_SECURE_LIB__ >= 200411L
```
shouldn't that rather be:
 `#if defined(_MSC_VER) && defined(__STDC_SECURE_LIB__) &&  __STDC_SECURE_LIB__ >= 200411L`
?

dra27 replied:
I can't find a place either for GCC or MSVC where it's explicitly documented, but an undefined token is treated as zero (this certainly works for the relevant old versions of cl)

oandrieu replied:
Didn't know that … it's in the C standard (§ 6.10.1):

> After all replacements due to macro expansion and the defined unary operator have been performed, all remaining identifiers (including those lexically identical to keywords) are replaced with the pp-number 0

dra27 replied:
Ah - good find! I was going to say that in the absence of documentation confirming the behaviour, it would be better to change it!

dra27 committed:
Fix compilation using Visual Studio .NET 2002

Microsoft introduced the `LL` suffix for integer literals in Visual
Studio .NET 2003 - earlier versions use `i64`

dra27 committed:
Correct floating point on old MSVC

Visual Studio 6 and earlier have somewhat insane handling of comparisons
with nan values. Provide alternate (slower) versions of float comparison
functions using isnan rather than standardized comparison behaviour.

dra27 committed:
Provide workaround for missing _vscprintf

_vscprintf was added to Microsoft Visual C++ in .NET 2002. Provide an
implementation of it for older compilers.

dra27 committed:
Correct missing header #define's in old MSVC

dra27 committed:
Fix compiler limitations in old MSVC

Older Microsoft C compilers can't generate code for casting unsigned
__int64 to double and the __pragma directive is not available for
suppressing warnings.

dra27 commented:
Four more commits extending OCaml right the way back to Visual Studio 6.0! I don't have a copy of Visual Studio 5.0, but I did test with Visual C++ 4 (!) which is unable to correctly parse the Windows 2003 Platform SDK headers, so for now Visual Studio 6 (1998) remains the oldest! I'm in the process of setting up a CI server for all of these versions, so the changes for now would not be abandoned. The patch for _vscprintf may well get adapted further - I think permitting that code to be used in the mingw port will re-enable Windows 2000 "support".

gasche commented:
cc @damiendoligez @xavierleroy ?

I would be in favor of merging these, but there is a mild risk of Windows regression so I wonder if Damien would want it in the 4.03 release.

dra27 commented:
I'm still running a lot of tests to verify these patches...

dra27 commented:
... and those tests were quite important because the floating point code had introduced a devastating regression in byterun/floats.c :blush:

damiendoligez commented on byterun/floats.c:
```diff
@@ -553,34 +553,41 @@ CAMLprim value caml_copysign_float(value f, value g)
+#ifdef LACKS_SANE_NAN
 
-CAMLprim value caml_neq_float(value f, value g)
+CAMLprim caml_neq_float(value vf, value vg)
```
Why did the return type disappear??

dra27 replied:
Ugh - a refactoring typo, I expect. It's amazing (by which I mean horrifying!) that it still works!

damiendoligez commented:
This is interesting, but I'm not feeling much pressure to integrate it before 4.03. On the other hand, I'm a bit wary of possible regressions.

Let's merge this into trunk just after branching 4.03.

dra27 commented:
You should be wary - 6573528 has introduced an msvc64 regression :rage: (which I shall shortly push a fix for).
It'll be great to see this on trunk and then use CI to track when supporting these old compilers really does become too much burden!
Is there a fixed date for branching 4.03?

alainfrisch commented:
@dra27 Out of curiosity:  which of older versions of MSVC have free versions available?

dra27 commented:
Visual Studio 2005 is the first one, I think - although there are various compilers and assemblers in older Windows SDKs, they were by MSDN subscription only, so not free. Added to that is that Microsoft only started including an x86 compiler in the Windows SDK with Windows Vista IIRC. I've got copies of VS 2002 & 2003 Pro thanks the Microsoft e-Academy licensing from when I was at Cambridge, and I bought VS 6 very cheaply shortly after VS 2002 came out. What was remarkably is the number of files still available from microsoft.com if you can determine the full URL! I managed to get 90/180 day trial of all versions from Visual Studio 2005+ onwards.
I haven't yet written it up properly, but there's a lot of info (including download URLs) on the page "Visual Studio Testing" at http://1drv.ms/1QeDkV0). There are free versions of both x64 and x86 available for Visual Studio 2008 onwards - the only free version of Visual Studio 2005 x64 (at least that I can find) is a pre-release version included in the Windows Server 2003 SP1 SDK.

alainfrisch commented:
While I really appreciate the effort, I'm wondering whether the added complexity is worth it.  @dra27 : do you have an actual need for supporting older MS C compilers, or is this work just a (fun) challenge?

dra27 commented:
(As it happens, I am still forced to use Visual Studio 6 for one client, but not for anything OCaml-related!!)
It was indeed as a challenge to see how far back it goes - what surprised me is that the complexity is fairly minimal and the significant parts affect only otherwise highly stable parts of the runtime. The blocks are all clearly delineated with #ifdef's showing that they're specific to very old compilers.
Assuming I can get CI running for this, would it be OK to put in trunk (post 4.03) so that support for them gets dropped when it's actually too much hassle (new feature requiring special support) rather than because they're old?

alainfrisch commented:
Thanks!  Rebased and merged on trunk.

dra27 commented:
Ta - there will be a few others when I get back to it ... the symlink stuff requires a few tweaks in the headers, but I'll open a new PR, obviously
